Add 3.14t wheel builds by skipping PyLIEF install

ngoldbaum opened the pull request:
Adds 3.14t to the wheel building matrix. Because PyLIEF doesn't yet ship cp314t wheels (https://github.com/lief-project/LIEF/issues/1255) and PyLIEF doesn't list any build requirements in its pyproject.toml, it's necessary to set up the PyLIEF build manually.

Unfortunately this adds a decent amount of overhead to the builds, particularly for the Mac builds, where the PyLIEF build happens twice during the build and install steps. We could probably optimize that but I'm curious if the maintainers are opposed to this approach and would prefer I work on getting PyLIEF wheels ready instead.

swap357 commented:
> Wow! Thanks for the context. That massively simplifies things.
> 
> While looking this over, I noticed that the `validate` steps are all "broken" on 3.14t - they install a Python 3.12 environment:
> 
> https://github.com/swap357/llvmlite/actions/runs/19111128292/job/54608653224?pr=126#step:3:521
> 
> That said, it may not actually matter that the validation step runs on Python 3.12, since the test is using LIEF to parse binaries, it may not be sensitive to the host or target Python version.



> Wow! Thanks for the context. That massively simplifies things.
> 
> While looking this over, I noticed that the `validate` steps are all "broken" on 3.14t - they install a Python 3.12 environment:
> 
> https://github.com/swap357/llvmlite/actions/runs/19111128292/job/54608653224?pr=126#step:3:521
> 
> That said, it may not actually matter that the validation step runs on Python 3.12, since the test is using LIEF to parse binaries, it may not be sensitive to the host or target Python version.

The validate step is due for refactoring. The step does 2 things -
validate imports with PyLIEF and `twine check` built wheel. 
The validate imports check there needs to be removed, it's redundant. It was added before we had imports test on `test_binding`, so that needs to be removed but I can do that on a separate PR.
For this PR, keeping validate as is would be okay. As you noted, it uses Python3.12 and won't have any problems.

ngoldbaum commented on .github/workflows/llvmlite_osx-arm64_wheel_builder.yml:
```diff
@@ -181,10 +186,13 @@ jobs:
           PYTHON_PATH=$(which "python${{ matrix.python-version }}")
           "$PYTHON_PATH" -m venv .venv && source .venv/bin/activate
-          python -m pip install --upgrade pip wheel
+          python -m pip install --upgrade pip setuptools wheel
```
leftover from an earlier attempt

```suggestion
          python -m pip install --upgrade pip wheel
```

ngoldbaum commented on .github/workflows/llvmlite_osx-arm64_wheel_builder.yml:
```diff
@@ -181,10 +186,13 @@ jobs:
+          if [[ "${{ matrix.python-version }}" != *t ]]; then
+              python -m pip install -v lief
+          fi
+          cd dist && python -m pip install -v ./llvmlite*.whl
```
```suggestion
          cd dist && python -m pip install -v ./*.whl
```

also leftover

ngoldbaum commented:
I rebased and squashed everything. I think this is ready now. 

Thanks for the pointer about it being a test-only dependency, I'm not sure where I got it in my head that it was a build dependency too.

ngoldbaum commented:
> Lastly are we all in agreement that this PR shows that llvmlite is very, very likely to work just fine ™️ in the free-threading/no-gil context?

If someone is using llvmlite in a single-threaded context or an effectively single-threaded context (e.g. no sharing state inside llvmlite between threads), then I think everything is OK. I do see a few issues when I do some multithreaded testing (see below) but I can also trigger all the issues I find on the GIL-enabled build as well, so they're not new.

When I try to run the llvmlite test suite using [unittest-ft](https://github.com/amyreese/unittest-ft) on both the GIL-enabled and free-threaded builds I see some test failures. Note that to actually run `unittest-ft` with `llvmlite`, the test suite needs a `load_tests` implementation and another small fix to `test_ir.py`:

```diff
diff --git a/llvmlite/tests/__init__.py b/llvmlite/tests/__init__.py
index 7f2b3b0..105b47e 100644
--- a/llvmlite/tests/__init__.py
+++ b/llvmlite/tests/__init__.py
@@ -1,4 +1,6 @@
+import os
 import sys
+import warnings

 import unittest
 from unittest import TestCase
@@ -27,6 +29,14 @@ def discover_tests(startdir):
     return suite


+def load_tests(loader, standard_tests, pattern):
+    # top level directory cached on loader instance
+    this_dir = os.path.dirname(__file__)
+    package_tests = loader.discover(start_dir=this_dir, pattern="test_*.py")
+    standard_tests.addTests(package_tests)
+    return standard_tests
+
+
 def run_tests(suite=None, xmloutput=None, verbosity=1):
     """
     args
diff --git a/llvmlite/tests/test_ir.py b/llvmlite/tests/test_ir.py
index d51f643..30cf26d 100644
--- a/llvmlite/tests/test_ir.py
+++ b/llvmlite/tests/test_ir.py
@@ -9,7 +9,8 @@ import re
 import textwrap
 import unittest

-from . import TestCase
+from unittest import TestCase
+
 from llvmlite import ir
 from llvmlite import binding as llvm
 from llvmlite import ir_layer_typed_pointers_enabled
@@ -548,49 +549,6 @@ class TestIR(TestBase):
```

`unittest-ft` runs every test in the test suite in a thread pool. It's not unusual for test suites to not be designed to be run like this and to be implemented using global state and this may be the case here. Here's some example failures I see running unittest-ft in stress-test mode (passing `-s`) and with randomized test order (passing `-r`):

```
==================================================================================
FAIL: test_fanout_3_limited (test_refprune.TestFanout.test_fanout_3_limited) (x1)
----------------------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/goldbaum/.pyenv/versions/3.13.7/lib/python3.13/site-packages/llvmlite/tests/test_refprune.py", line 434, in test_fanout_3_limited
    self.assertEqual(stats.fanout, 0)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
AssertionError: 3 != 0

==================================================================================
FAIL: test_fanout_1 (test_refprune.TestFanout.test_fanout_1) (x1)
----------------------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/goldbaum/.pyenv/versions/3.13.7/lib/python3.13/site-packages/llvmlite/tests/test_refprune.py", line 388, in test_fanout_1
    self.assertEqual(stats.fanout, 3)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
AssertionError: 6 != 3

==================================================================================
FAIL: test_fanout_raise_2 (test_refprune.TestFanoutRaise.test_fanout_raise_2) (x1)
----------------------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/goldbaum/.pyenv/versions/3.13.7/lib/python3.13/site-packages/llvmlite/tests/test_refprune.py", line 479, in test_fanout_raise_2
    self.assertEqual(stats.fanout_raise, 0)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: 2 != 0

==================================================================================
FAIL: test_per_diamond_4 (test_refprune.TestDiamond.test_per_diamond_4) (x1)
----------------------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/goldbaum/.pyenv/versions/3.13.7/lib/python3.13/site-packages/llvmlite/tests/test_refprune.py", line 342, in test_per_diamond_4
    self.assertEqual(stats.diamond, 2)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^
AssertionError: 4 != 2

==================================================================================
FAIL: test_per_diamond_3 (test_refprune.TestDiamond.test_per_diamond_3) (x1)
----------------------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/goldbaum/.pyenv/versions/3.13.7/lib/python3.13/site-packages/llvmlite/tests/test_refprune.py", line 322, in test_per_diamond_3
    self.assertEqual(stats.diamond, 0)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^
AssertionError: 2 != 0
----------------------------------------------------------------------------------
```

Because this is a multithreaded test of code that exhibits a race condition, all of these failures are nondeterministic. I don't immediately see where the race is happening. Given this happens on the GIL-enabled build and the free-threaded build, I don't think you should worry about it now with any urgency because you're adding support for the free-threaded build. Probably worth investigating and understanding, though.

I also see an explicitly multithreaded test in `test_binding.py`, which passes on the free-threaded. There's only the one test and adding more tests of supported multithreaded workflows can't hurt. It'd also probably help to add some docs on what is and isn't supported in a multithreaded environment.

Another thing that might be worth doing is setting up a [Thread Sanitizer build](https://py-free-threading.github.io/thread_sanitizer/) and then run the existing multithreaded test and `unittest-ft` to check for data races. That said, I think that might require a TSan build of LLVM itself, which might be a tall order.

ngoldbaum commented:
> Thanks for the patch you provided. It made reproducing the unittest-ft errors easier. It looks like the stats tracking on ref prune pass isn't thread safe. If the tests are run in parallel, counts would be messed because all threads would update same global counters. I think we'd need to make those thread_local to track its own stats so they stay isolated.

Makes sense! unittest-ft doesn't yet have a feature to mark tests as thread-unsafe but I think it needs it for cases like this: https://github.com/amyreese/unittest-ft/issues/42.

In any case, happy to hear you think this PR is ready!

swap357 commented:
> @swap357 thank you for the review here. Can you create a followup issue to remove the `pylief` skip again when 3.14t wheels for that package become available?

yeah, tracking issue for `pylief`- https://github.com/numba/llvmlite/issues/1388
